### ISSUE TITLE
Deeper access to internal AJV?

### ISSUE THREAD

**Issue by dougrich** [fastify/fastify#520]
Looking at #493 it seems there's a precedent for running into this.

Some specific scenarios that I can't accomplish without deeper access today are:

- shared reference schemas
  - i.e., [this gist](https://gist.github.com/dougrich/d104baa418b88582c85a8b02f49bb6c6) shows a schema with shared references and how we can use both AJV and fast-json-stringify with it
  - I can achieve this today by manually resolving references before passing it into fastify
- re-used schemas
  - schemas that are re-used seem to be recompiled
  - I can achieve this today be creating a second AJV instance and providing our own `schemaCompiler` that provides cached schemas

The closest I can get is to create our own AJV, separate to the one instantiated in fastify, and provide our own `schemaCompiler` that does a `getSchema`, and serializing manually using our own schemas compiled with `fast-json-stringify`.

What would be ideal is:

```js
const fastify = require('fastify')();

fastify.addSchema(require('../schema/api.json')

const opts = {
  schemaId: 'api.json#/definitions/instance:get'
}

// under the hood this does a getSchema call
fastify.get('/case1', opts, function (req, reply) {
  ...
})

// since this is also doing a getSchema call, it re-uses the existing compiled schema in AJV
fastify.get('/case2', opts, function (req, reply) {
  ...
})

// this schema uses the same definitions as the instance:get schema
const putOpts = {
  schemaId: 'api.json#/definitions/instance:upsert'
}

fastify.put('/case1', putOpts, function (req, reply) {
  ...
})
```

Alternatively, this could be a third party library, but it feels like it would duplicate a large part of fastify's validation and serialization logic.

Thoughts?

P.S.

In your [code of conduct](https://github.com/fastify/fastify/blob/master/CODE_OF_CONDUCT.md) you've still got an [INSERT EMAIL ADDRESS] placeholder in the Enforcement section.

I'm also a big fan of fastify - I really like the focus on performance. Thank you for all your hard work on this.

**Comment by mcollina** [comment]
I think we need a better story on how to reuse schemas across the library.
Would you like to send a PR?

**Comment by delvedor** [comment]
If you want to have different serialization schemas you can also use [`any-schema-you-like`](https://github.com/fastify/any-schema-you-like) :)

**Comment by zenflow** [comment]
Yes, there is some power and features of Ajv that we can't use, without adding `ajv` to our package dependencies and using `fastify.setSchemaCompiler` and doing some things ourselves, which isn't the end of the world, but it would be nice if we could avoid it.

Besides `addSchema` functionality, I think the [`addKeyword` functionality](https://github.com/epoberezkin/ajv#addkeywordstring-keyword-object-definition---ajv) is also very important, if you are validating content and not just shape/structure. It is not currently possible to use Ajv's support for custom keywords without adding `ajv` to our package dependencies, etc..

Here is the API I proposed in PR #495 (which includes a test similar to this example):

```
const fastify = Fastify({
  ajv: (Ajv, defaultOptions) => {
    return new Ajv(defaultOptions)
      .addSchema(require('./example-schema-set.json'));
  }
})
```

(This in addition to accepting an object for the `ajv` option, of course)

That said, I'm guessing first-class support for `addSchema` (i.e. `fastify.addSchema`) would be way better than apps configuring Ajv in this way themselves directly.

**Comment by dougrich** [comment]
PR #528 is open with the API interface + behavior I proposed above.

I mentioned two open discussion points that need to be resolved before merging in: @zenflow's ask for the `addKeyword` functionality and an open question on if the public API should expose `addSchema` or if all schemas should be passed in through `options.ajv.schemas`.

I'm personally preferential to `addSchema` as that means it'd be possible to add schemas alongside routes, but I'm fine with either way in the long run.

I'm not sure what the best approach with `addKeyword` is, since I've not personally needed to use it. I see @zenflow's use case but looking back at the discussion in some of those previous issues I'm not sure how much of AJV's api you want to let flow through before just exposing `fastify.ajv` directly.

Please let me know if you have any feedback on the PR.

**Comment by jtlapp** [comment]
Another use case. I typically add an `xregexp` keyword to AJV to validate with [`XRegExp`](http://xregexp.com/), so that I get classes of Unicode characters, which the JSON Schema spec [does not permit](http://json-schema.org/latest/json-schema-validation.html#rfc.section.4.3).

I typically do this to ensure client strings are clean: `XRegExp("^[^\\p{C}\\p{Zl}\\p{Zp}]+$")`.

**Comment by nwoltman** [comment]
Closed by #674
